client/mysqltest.cc

```diff
@@ -3809,45 +3809,305 @@ void do_remove_files_wildcard(struct st_command *command)
+  static DYNAMIC_STRING ds_wild;
+  static DYNAMIC_STRING ds_recursive;
+  static bool recursive;
+  MY_STAT stat_buff;
   const struct command_arg copy_file_args[] = {
-    { "from_file", ARG_STRING, TRUE, &ds_from_file, "Filename to copy from" },
-    { "to_file", ARG_STRING, TRUE, &ds_to_file, "Filename to copy to" }
+    { "source", ARG_STRING, TRUE, &ds_source, "File/dir name to copy from" },
+    { "destination", ARG_STRING, TRUE, &ds_destination, "File/dir name to copy to" },
+    { "recursive", ARG_STRING, FALSE, &ds_recursive, "Optional argument indicating if directories should be copied recursively" }
   };
-  DBUG_ENTER("do_copy_file");
 
   check_command_args(command, command->first_argument,
                      copy_file_args,
                      sizeof(copy_file_args)/sizeof(struct command_arg),
                      ' ');
+  recursive = !strcmp("-r", ds_recursive.str);
 
-  if (bad_path(ds_to_file.str))
-    DBUG_VOID_RETURN;
+  /*
+    Throw an error if destination path is invalid, or source and destination paths are same.
+  */
+  if (bad_path(ds_destination.str) ||
+      !strcmp(ds_source.str, ds_destination.str))
+  {
+    error = 1;
+  }
+
+
+  else if (my_stat(ds_source.str, &stat_buff, MYF(0))) {
+
+    /* For regular file, copy directly */
+    if (MY_S_ISREG(stat_buff.st_mode)) {
+      DBUG_PRINT("info", ("Copy %s to %s", ds_source.str, ds_destination.str));
+      /* MY_HOLD_ORIGINAL_MODES prevents attempts to chown the file */
+      error= (my_copy(ds_source.str, ds_destination.str,
+                      MYF(MY_DONT_OVERWRITE_FILE | MY_WME | MY_HOLD_ORIGINAL_MODES)) != 0);
+    }
+
+    /* For directory, copy if recursive is enabled */
+    else if (recursive) {
+      DBUG_PRINT("info", ("Recursive copy files of %s to %s", ds_source.str, ds_destination.str));
+      DBUG_PRINT("info", ("listing directory: %s", ds_source.str));
+      error = recursive_copy(&ds_source, &ds_destination);
+    }
+
+    else {
+      error = 1;
+    }
+  }
+
+  else {
+    /*
+      Cannot get file stat. Could be a wildcard pattern. Extract the basename as wildcard pattern.
+      If it is not a vaild wildcard pattern, wildcard_copy() will return error.
+    */
+    const char *wild = my_basename(ds_source.str);
+    init_dynamic_string(&ds_wild, wild, strlen(wild) + 1, 0);
+    dynstr_trunc(&ds_source, ds_wild.length+1);
+    error = wildcard_copy(&ds_source, &ds_destination, &ds_wild, recursive);
+  }
 
-  DBUG_PRINT("info", ("Copy %s to %s", ds_from_file.str, ds_to_file.str));
-  /* MY_HOLD_ORIGINAL_MODES prevents attempts to chown the file */
-  error= (my_copy(ds_from_file.str, ds_to_file.str,
-                  MYF(MY_DONT_OVERWRITE_FILE | MY_WME | MY_HOLD_ORIGINAL_MODES)) != 0);
   handle_command_error(command, error, my_errno);
-  dynstr_free(&ds_from_file);
-  dynstr_free(&ds_to_file);
+  dynstr_free(&ds_source);
+  dynstr_free(&ds_destination);
+  dynstr_free(&ds_wild);
+  dynstr_free(&ds_recursive);
   DBUG_VOID_RETURN;
```

Comment:
This is the line where Windows says there is a failure, but I don't see why right now:

```
         D:\Buildbot\amd64-windows\build\client\mysqltest.cc(4111,3): error C2059: syntax error: '.' [D:\Buildbot\amd64-windows\build\client\mariadb-test.vcxproj]
       D:\Buildbot\amd64-windows\build\client\mysqltest.cc(4111,3): error C2059:   DBUG_VOID_RETURN; [D:\Buildbot\amd64-windows\build\client\mariadb-test.vcxproj]
       D:\Buildbot\amd64-windows\build\client\mysqltest.cc(4111,3): error C2059:   ^ [D:\Buildbot\amd64-windows\build\client\mariadb-test.vcxproj]
         D:\Buildbot\amd64-windows\build\client\mysqltest.cc(4111,3): error C2275: '_db_stack_frame_': illegal use of this type as an expression [D:\Buildbot\amd64-windows\build\client\mariadb-test.vcxproj]
       D:\Buildbot\amd64-windows\build\client\mysqltest.cc(4111,3): error C2275:   DBUG_VOID_RETURN; [D:\Buildbot\amd64-windows\build\client\mariadb-test.vcxproj]
       D:\Buildbot\amd64-windows\build\client\mysqltest.cc(4111,3): error C2275:   ^ [D:\Buildbot\amd64-windows\build\client\mariadb-test.vcxproj]
```